### PR TITLE
Improved support for time-dependent discrete loads

### PR DESCRIPTION
**Pull request by kmokstad**
This includes support for (time)derivatives of ScalarFunc objects and accounting to time-variation when calculating the external energy contributions. This is needed when we are going to do nonlinear shell and beam problems with varying point loads.

This partly supersedes one commit in #257 which will be rebased after this one.

### CODE REVIEW

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=46 please

File: src/Utility/ExprFunctions.C [hunk 1]
```diff
@@ -104,50 +106,59 @@ EvalFunc::EvalFunc (const char* function, const char* x)
...
+  for (ExprEval::Expression* it : expr)
     delete it;
-  for (auto& it : f)
+  for (ExprEval::FunctionList* it : f)
     delete it;
-  for (auto& it : v)
+  for (ExprEval::ValueList* it : v)
     delete it;
-  arg.clear();
+  if (gradient)
```
Comment (akva2) [inline src/Utility/ExprFunctions.C]: delete is null safe.

File: src/Utility/ExprFunctions.C [hunk 2]
```diff
@@ -205,26 +226,26 @@ EvalFunction::EvalFunction (const char* function) : gradient{}, dgradient{}
...
+  for (ExprEval::Expression* it : expr)
     delete it;
-  for (auto& it : f)
+  for (ExprEval::FunctionList* it : f)
     delete it;
-  for (auto& it : v)
+  for (ExprEval::ValueList* it : v)
     delete it;
-  for (auto& it : gradient)
+  for (EvalFunction* it : gradient)
```
Comment (akva2) [inline src/Utility/ExprFunctions.C]: clear these as well i guess.

File: src/Utility/ExprFunctions.h [hunk 3]
```diff
@@ -42,23 +42,36 @@ class EvalFunc : public ScalarFunc
...
+
+  //! \brief Returns the first-derivative of the function.
+  virtual Real deriv(Real x) const;
 
 protected:
   //! \brief Non-implemented copy constructor to disallow copying.
   EvalFunc(const EvalFunc&);
-  //! \brief Non-implemented assigment operator to disallow copying.
+  //! \brief Non-implemented assignment operator to disallow copying.
   EvalFunc& operator=(const EvalFunc&);
```
Comment (akva2) [inline src/Utility/ExprFunctions.h]: c++11 has a useful construct - = delete. yields more understandable error messages should this be attempted to be used.

File: src/Utility/ExprFunctions.C (hunk 1 above)
Comment (kmokstad) [reply src/Utility/ExprFunctions.C]: OK, removed the if ()

File: src/Utility/ExprFunctions.C (hunk 2 above)
Comment (kmokstad) [reply src/Utility/ExprFunctions.C]: not sure about that. gradient (and dgradient) are std::array's and not vectors. Then clear is not defined. But do we need to nullify the array members (pointers) or is that automatically done through the delete operator.

File: src/Utility/ExprFunctions.h (hunk 3 above)
Comment (kmokstad) [reply src/Utility/ExprFunctions.h]: changed.

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=46 please

File: src/Utility/ExprFunctions.C (hunk 2 above)
Comment (akva2) [reply src/Utility/ExprFunctions.C]: delete does not nullify, std::fill or for() then.

**Comment by akva2** [comment]
jenkins does not use the Apps/ CMakeLists.txt, so errors there will slip.

File: src/Utility/ExprFunctions.C (hunk 2 above)
Comment (kmokstad) [reply src/Utility/ExprFunctions.C]: std::array has a fill member, so used that instead.

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=46 please

**Pull request merged**